### DISCUSSION TITLE
Ensuring sensitive code doesn't leak to the client side bundle

### DISCUSSION THREAD

**Discussion by benoitv-code** [vercel/next.js#41846] [Help] [answered]
Hi!

# Context

**I am trying to figure out the steps that I need to take to ensure secrets are not leaked client-side in the bundle generated by Next.js**

By experimenting with the next-code-elimination tool, I noticed that a lot more variables are included that I would have expected (see the different examples): https://next-code-elimination.vercel.app/s/SHXYoSymF

See the output (either in the link or in the screenshot below):
- Unused local variables are included in the client side bundle
- All local config keys/values are included in the client side bundle if one key is used by the component
- All exported variables are included in the client side bundle

# Questions

Based on these examples, I understand that storing things like secret seeds directly in a `const config = {};` object in a `./helpers/config.ts` file (for example) is a bad idea if the file ends up being included by a page component (even if the secret keys are never used client-side).

I also understand that exported variables might leak if a helper file ends up being included in the bundle?

Based on this, I understand that the safest way to include such secrets or sensitive configurations would be to do it through an `.env` file.

**Are the above assumptions correct, or is the risk of leak lower than I assume?**

_Note: I know that "it's bad to store secrets in code", but it happens that some configurations (like seeds, or limits) can be sensitive too, hence the question to understand what can actually end up being bundled client-side_

Thanks for the answers!

# Screenshot
![2022-10-26 12_14_02-](https://user-images.githubusercontent.com/471168/198000636-1c0ccb35-6f2f-4774-b896-1f95afacd80f.png)

**Comment by icyJoseph** [comment, marked as answer]
I'd say, don't play with fire, use `.env` (and company) which then Next.js uses to populate process.env - Variables prefixed with `NEXT_PUBLIC_` will be included in the client side bundle, inlined, and the rest needs to be actually be invoked, oversimplifying a bit here.

Everything you consider to be "sensitive" should be loaded through those, unless you are willing to pay the price of investigating if there's a leak every single time a code change is made.

In the example you give, the problem is that the elimination tool cannot figure out if this objects will be mutated, or accessed at runtime, or rather, where the access will happen, if ever. If they were used within data fetching functions (GSSP for example), then the elimination tool drops them from the client.

Though in this example:

```jsx
const usersPerPage = config_local_one_field_client_side.usersPerPage
export default function ({ }) {
  console.log(usersPerPage);
  console.log(usersPerPage_local_client_side);

  return (
    <Layout />
  );
}
```
 
We might think about the bundle dropping `config_local_one_field_client_side`, but mutations are a thing, and since `usersPerPage` is copied by value, perhaps we'll introduce a bug if we replace it at build time by it's initial value.

Everything you export from a page that's not known to Next.js is also kept by webpack (like secretSeed_exported_server_side). 

You can also tell webpack to drop certain imports depending on the runtime we bundle for.

There's also the issue that dynamic imports might also end up leaking data in spite of tree-shaking, unless you command the tree-shaking of dynamic imports through magic comments.

**Reply by benoitv-code** [reply]
Providing an update to the above answer:

Next.js 13 was just published, and they provide this additional tips in their doc:

> To prevent this sort of unintended client usage of server code, we can use the server-only package to give other developers a build-time error if they ever accidentally import one of these modules into a Client Component.
> 
> ```typescript
> import "server-only";
> 
> export async function getData() {
>   let resp = await fetch("https://external-service.com/data", {
>     headers: {
>       authorization: process.env.API_KEY,
>     },
>   });
> 
>   return resp.json();
> }
> ```

Source: https://beta.nextjs.org/docs/rendering/server-and-client-components#keeping-server-only-code-out-of-client-components-poisoning

_Note: Only seems to work for code already migrated to the new `app/` structure_
